presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java

```diff
@@ -0,0 +1,227 @@
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.PARTIAL;
+import static com.facebook.presto.sql.planner.plan.AggregationNode.Step.SINGLE;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.GATHER;
+import static com.facebook.presto.sql.planner.plan.ExchangeNode.Type.REPARTITION;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+
+public class PushPartialAggregationThroughExchange
+        implements Rule
+{
+    private static final Pattern PATTERN = Pattern.typeOf(AggregationNode.class);
+
+    private final FunctionRegistry functionRegistry;
+
+    public PushPartialAggregationThroughExchange(FunctionRegistry functionRegistry)
+    {
+        this.functionRegistry = functionRegistry;
+    }
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Optional<PlanNode> apply(PlanNode node, Context context)
+    {
+        AggregationNode aggregationNode = (AggregationNode) node;
+
+        PlanNode childNode = context.getLookup().resolve(aggregationNode.getSource());
+        if (!(childNode instanceof ExchangeNode)) {
+            return Optional.empty();
+        }
+
+        ExchangeNode exchangeNode = (ExchangeNode) childNode;
+
+        boolean decomposable = aggregationNode.isDecomposable(functionRegistry);
+
+        if (aggregationNode.getStep().equals(SINGLE) &&
+                aggregationNode.hasEmptyGroupingSet() &&
+                aggregationNode.hasNonEmptyGroupingSet()) {
+            checkState(
+                    decomposable,
+                    "Distributed aggregation with empty grouping set requires partial but functions are not decomposable");
+            return Optional.of(split(aggregationNode, context));
+        }
+
+        if (!decomposable) {
+            return Optional.empty();
+        }
+
+        // partial aggregation can only be pushed through exchange that doesn't change
+        // the cardinality of the stream (i.e., gather or repartition)
+        if ((exchangeNode.getType() != GATHER && exchangeNode.getType() != REPARTITION) ||
+                exchangeNode.getPartitioningScheme().isReplicateNullsAndAny()) {
+            return Optional.empty();
+        }
+
+        if (exchangeNode.getType() == REPARTITION) {
+            // if partitioning columns are not a subset of grouping keys,
+            // we can't push this through
+            List<Symbol> partitioningColumns = exchangeNode.getPartitioningScheme()
+                    .getPartitioning()
+                    .getArguments()
+                    .stream()
+                    .filter(Partitioning.ArgumentBinding::isVariable)
+                    .map(Partitioning.ArgumentBinding::getColumn)
+                    .collect(Collectors.toList());
+
+            if (!aggregationNode.getGroupingKeys().containsAll(partitioningColumns)) {
```

Comment:
Does it matter from which grouping set symbols come? 

For example with partition on `a, b` and grouping sets `(a,c), (a, b)` you condition will be satisfied. Is it ok?